### DISCUSSION TITLE
Download video with yt-dlp

### DISCUSSION THREAD

**Discussion by RicardoBassete** [johnlindquist/kit#1499] [Share]
[Open download-with-yd-dlp in Script Kit](https://scriptkit.com/api/new?name=download-with-yd-dlp&url=https://gist.githubusercontent.com/RicardoBassete/ed260621be7c8ff5eb215445a1d57488/raw/24e7880e430e37291e0b1545aef02b38a560669a/download-with-yd-dlp.ts")

```js
// Name: Download video with yt-dlp
// Description: Download video with yt-dlp
// Author: Ricardo Gonçalves Bassete

import "@johnlindquist/kit"

const outDir = home('Downloads', 'yt-dlp')

const flags = {
  mp4: `-f "bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best" -o "%(title)s.mp4"`,
  mp3: `-o "%(title)s.mp3" -x --audio-format mp3 --audio-quality 0`,
}

const videoURl = await arg({
  placeholder: 'Video URL',
  alwaysOnTop: true
})

const format = await arg({
  placeholder: 'Select format',
  alwaysOnTop: true,
  choices: Object.keys(flags),
  strict: true
})

await hide()

await ensureDir(outDir)
cd(outDir)
await exec(`yt-dlp ${flags[format]} "${videoURl}"`)
notify({
  title: 'Finished downloading video',
  message: `Location: ${outDir}`
})
```

**Comment by kenNash** [comment]
Hi @RicardoBassete, thank you for your contribution!

When I use your script (macOS 14.5), two files are downloaded: one audio (only) and one video (only).
Is it correct? Any chance to have one single file with audio+video?

Thanks in advance

**Reply by RicardoBassete** [reply]
That's strange, It's supposed to download just one file, I'm on Windows 11, and it downloads only one file with audio and video. 
Maybe if you use the `$` function instead of `exec` it may help, but that is kind of a shot in the dark.
So instead of using:
```js
await exec(`yt-dlp ${flags[format]} "${videoURl}"`)
```
You may use something like this:
```js
await $`yt-dlp ${flags[format]} "${videoURl}"`
```
If even that doesn't work, you can try using `ffmpeg` to combine the audio and video files into one and then remove the old files.
Something like this:
```js
await $`ffmpeg -i "videoFile.mp4" -i "audioFile.mp3" -shortest "outPutFile.mp4"`
awair remove("videoFile.mp4")
awair remove("audioFile.mp3")
```
I hope this can help you🙂

**Comment by kenNash** [comment]
Thank you @RicardoBassete for your prompt answer!

Actually the `$` version throws me this error:
```
pieces.some is not a function
TypeError: pieces.some is not a function
    at Proxy.set (file:///Users/kennash/.kit/node_modules/zx/build/core.js:59:16)
    at global.<computed> (file:///Users/kennash/.kit/target/app.js:1734:31)
    at file:///Users/kennash/.kenv/scripts/download-video-with-yt-dlp.ts?now=1720826633350.kit:21:7
```
(which I don't understand at all…).

Using the `exec` version on `https://www.youtube.com/watch?v=YlW1XhbWHeI`, e.g., and selecting `mp4` at the `flags` choice, these two files are downoaded:
- The Heart Asks Pleasure First (from The Piano) - Analysis and Performance.f137.mp4
- The Heart Asks Pleasure First (from The Piano) - Analysis and Performance.f140.m4a

I'll give the ffmpeg solution a try.

Again, thank you. 🙏
